src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDependencyDao.java

```diff
@@ -67,9 +68,19 @@ public boolean datasetHasSnapshotReference(Dataset dataset) {
         FireStoreProject fireStoreProject = FireStoreProject.get(dataset.getProjectResource().getGoogleProjectId());
         String dependencyCollectionName = getDatasetDependencyId(dataset.getId().toString());
         CollectionReference depColl = fireStoreProject.getFirestore().collection(dependencyCollectionName);
-        // check to see if the datasets collection contains any dependencies
-        boolean hasDependencies = depColl.listDocuments().iterator().hasNext();
-        return hasDependencies;
+        // TODO - use generic retry library
+        int retryCount = 0;
+        while (true) {
+            try {
+                // check to see if the datasets collection contains any dependencies
+                return depColl.listDocuments().iterator().hasNext();
+            } catch (FirestoreException ex) {
+                retryCount++;
+                if (retryCount > fireStoreUtils.getFirestoreRetries()) {
+                    throw ex;
```

Comment:
Could we add a message to the thrown exception here to note that the operation was retried `retryCount` times? I see that  `FirestoreException` doesn't have a constructor allowing us to specify a causal exception, so we can't just wrap it in another `FirestoreException`. Maybe a log message would suffice instead of throwing a different exception if its too much effort.

Reply:
Could we include the number of retries in that message? It would be a useful smoke indicator as to what might be going on.